sound/soc/sof/ipc4-pcm.c

```diff
@@ -32,6 +35,108 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
 }
 EXPORT_SYMBOL(sof_ipc4_set_pipeline_state);
 
+static int sof_ipc4_build_time_info(struct snd_sof_dev *sdev, struct snd_sof_pcm_stream *spcm)
+{
+	struct sof_ipc4_pipeline_registers ppl_reg;
+	struct sof_ipc4_llp_reading_slot llp_slot;
+	struct sof_ipc4_timestamp_info *info;
+	struct sof_ipc4_copier *host_copier = NULL;
+	struct sof_ipc4_copier *dai_copier = NULL;
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_dai *dai;
+	u32 node_id;
+	u32 offset;
+	int i;
+
+	/* find host & dai to locate info in memory window */
+	for_each_dapm_widgets(spcm->list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+
+		if (!swidget)
+			continue;
+
+		if (WIDGET_IS_AIF(swidget->widget->id)) {
+			host_copier = swidget->private;
+		} else if (WIDGET_IS_DAI(swidget->widget->id)) {
+			dai = swidget->private;
+			dai_copier = dai->private;
+		}
+	}
+
+	info = spcm->time_info;
+
+	if (!host_copier || !dai_copier) {
+		dev_err(sdev->dev, "failed to find host copier %p or dai copier%p\n",
+			host_copier, dai_copier);
```

Comment:
do we need the %p ? At least one of the two is going to be NULL.

And if we do there's a missing space in 'dai copier%p'

Reply:
I will refine it. Thanks!